### PR TITLE
prep zap for fee exclusion:

### PR DESCRIPTION
**Pull request by 0xCaps**
- force max bond for a given number of nfts
- policy control for minimum % of an nft that can be bonded with fee exclusion

### CODE REVIEW

File: contracts/interfaces/zaps/IMintAndBond.sol [hunk 1]
```diff
@@ -1,13 +1,13 @@
 // SPDX-License-Identifier: MIT
 pragma solidity ^0.7.5;
 
+import "../IERC20.sol";
```
Comment (0xCaps) [inline contracts/interfaces/zaps/IMintAndBond.sol]: can remove this

Comment (0xNom) [reply contracts/interfaces/zaps/IMintAndBond.sol]: Was just about to comment this 😁

File: contracts/zaps/MintAndBond.sol [hunk 2]
```diff
@@ -87,4 +103,25 @@ contract MintAndBond is IMintAndBond, ReentrancyGuard {
     return vault;
   }
 
+  function _getMaxPayout(uint256 bondId) internal returns (uint64 maxPayout) {
+    (,,,,maxPayout,,) = bondDepository.markets(bondId);
+  }
+
+  /**
+   * @notice                minimum fraction of an NFT that can be bonded
```
Comment (0xNom) [inline contracts/zaps/MintAndBond.sol]: Quite a lot of spaces after `@notice`, is this needed?

File: contracts/zaps/MintAndBond.sol [hunk 3]
```diff
@@ -87,4 +103,25 @@ contract MintAndBond is IMintAndBond, ReentrancyGuard {
     return vault;
   }
 
+  function _getMaxPayout(uint256 bondId) internal returns (uint64 maxPayout) {
+    (,,,,maxPayout,,) = bondDepository.markets(bondId);
+  }
```
Comment (0xNom) [inline contracts/zaps/MintAndBond.sol]: Do we need this as a separate method call? Could probably have this inline

File: contracts/zaps/MintAndBond.sol [hunk 4]
```diff
@@ -87,4 +103,25 @@ contract MintAndBond is IMintAndBond, ReentrancyGuard {
...
+   * @notice                minimum fraction of an NFT that can be bonded
+   * @param _minBondAmount  amount in 18 decimal
+   */
+  function setMinBondAmount(uint256 _minBondAmount) external onlyPolicy {
+    require(_minBondAmount > 0, "Amount not set");
+    require(_minBondAmount < 1e18, "Amount too high");
+    minBondAmount = _minBondAmount;
+  }
+
+  function rescue(address token) external onlyGovernor {
```
Comment (0xNom) [inline contracts/zaps/MintAndBond.sol]: Do we expect that tokens may get stuck in this flow or is this more of a "just-in-case"?

File: scripts/deployTokenWethCalculator.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
+const {ethers} = require("hardhat");
+const token = '0x87931E7AD81914e7898d07c68F145fC0A553D8Fb';
```
Comment (0xNom) [inline scripts/deployTokenWethCalculator.js]: Can we comment what token this is

File: contracts/zaps/MintAndBond.sol (hunk 4 above)
Comment (0xCaps) [reply contracts/zaps/MintAndBond.sol]: Just in-case, contracts not upgradeable so it's nice protection to have.

File: contracts/zaps/MintAndBond.sol [hunk 6]
```diff
@@ -5,69 +5,159 @@ import "../interfaces/IBondDepository.sol";
...
+   * @notice             claim notes for user
+   * @param _user        the user to claim for
+   * @param _indexes     the note indexes to claim
+   * @return amount_     sum of amount sent in vToken
+   */
+  function claim(address _user, uint256[] memory _indexes, address _vault) external override returns (uint256 amount_) {
+    uint48 time = uint48(block.timestamp);
+
+    for (uint256 i = 0; i < _indexes.length; i++) {
+      (uint256 pay, bool matured) = pendingFor(_user, _indexes[i], _vault);
```
Comment (tomwade) [inline contracts/zaps/MintAndBond.sol:110]: This should save gas

uint256 indexes = _indexes.length;
for (uint256 i; i < indexes;) {
  ...
  unchecked { ++i; }
}

File: contracts/zaps/MintAndBond.sol [hunk 7]
```diff
@@ -5,69 +5,159 @@ import "../interfaces/IBondDepository.sol";
...
+      require(matured, "Depository: note not matured");
```
Comment (tomwade) [inline contracts/zaps/MintAndBond.sol:111]: Do we want to require here? Seems like the if statement after would skip any that cannot be claimed

File: contracts/zaps/MintAndBond.sol [hunk 8]
```diff
@@ -5,69 +5,159 @@ import "../interfaces/IBondDepository.sol";
...
+        notes[_user][_vault][_indexes[i]].claimed = time; // mark as claimed
+        amount_ += pay;
+      }
+    }
 
-    // Return remaining ERC20
-    uint256 remaining = vaultToken.balanceOf(address(this));
-    vaultToken.safeTransfer(to, remaining);
+    IERC20 vaultToken = IERC20(_vault);
+    vaultToken.safeTransfer(_user, amount_);
```
Comment (tomwade) [inline contracts/zaps/MintAndBond.sol:119]: Do we need to protect against Reentrancy here as doing a safe transfer to the User could trigger an additional claim etc?

**Review by tomwade — APPROVED** [review]

**Pull request merged**